### PR TITLE
Read stage0 dice data in restricted kernel

### PR DESCRIPTION
**Pull request by jul-sh**
Please carefully review. I arrived at this code after some experimentation. It contains unsafe code. I do not claim fully understand the nuances of memory management here, though I am keen to learn. 

Left some questions as comments.

### CODE REVIEW

File: oak_restricted_kernel/src/lib.rs [hunk 1]
```diff
@@ -156,6 +156,36 @@ pub fn start_kernel(info: &BootParams) -> ! {
...
+        dice_memory_slice.fill(0);
+
+        if dice_data.magic != oak_dice::evidence::STAGE0_MAGIC {
+            panic!("dice data loaded from stage0 failed validation");
+        }
+        dice_data
+    };
+
     // Physical frame allocator
     mm::init(info.e820_table(), program_headers);
```
Comment (jul-sh) [inline oak_restricted_kernel/src/lib.rs:190]: implementing this code after page tables are initialized leads to page faults, unsure why.

File: oak_restricted_kernel/src/lib.rs [hunk 2]
```diff
@@ -156,6 +156,36 @@ pub fn start_kernel(info: &BootParams) -> ! {
...
     let program_headers = unsafe { elf::get_phdrs(VirtAddr::new(0x20_0000)) };
 
+    let _stage0_dice_data = {
+        let dice_memory_slice = {
+            let e820_dice_data_entry = info
+                .e820_table()
+                .iter()
+                .find(|e| e.entry_type() == Some(oak_linux_boot_params::E820EntryType::DiceData))
+                .expect("failed to find dice data");
+
```
Comment (jul-sh) [inline oak_restricted_kernel/src/lib.rs:166]: initially I had another section here that validated that the E820 table matched any kernel args that are presented (nice to have, linux it them but restricted kernel could ignore the relevant kernel arg)

```
  // Validate that the dice data mem address matches the kernel args if present
  if let Some(kernel_arg_dice_addr) =
      kernel_args.get(oak_dice::evidence::DICE_DATA_CMDLINE_PARAM)
  {
      let parsed_addr =
          usize::from_str(kernel_arg_dice_addr).expect("failed to parse oak dice flag");
      if parsed_addr != e820_dice_data_entry.addr() {
          panic!("inconsistent dice data addresses supplied in the E820 and kernel args")
      }
  }
```

however, this led to memory issues, unsure why.

**Review by jul-sh — COMMENTED** [review]
left unanswered questions I had about the inner workings about memory in this part of the kernel. Responses are appreciated to deepen my understanding.

**Review by conradgrobler — COMMENTED** [review]
Thanks

File: oak_restricted_kernel/src/lib.rs [hunk 3]
```diff
@@ -156,6 +156,36 @@ pub fn start_kernel(info: &BootParams) -> ! {
...
+            // safety: the E820 table indicated that this is the corrct memory segment
```
[the remark this answers is not in the document]
Comment (jul-sh) [reply oak_restricted_kernel/src/lib.rs:167]: done, thanks!

File: oak_restricted_kernel/src/lib.rs [hunk 4]
```diff
@@ -156,6 +156,36 @@ pub fn start_kernel(info: &BootParams) -> ! {
...
+            unsafe {
+                core::slice::from_raw_parts_mut::<u8>(
+                    e820_dice_data_entry.addr() as *mut u8,
+                    e820_dice_data_entry.size(),
+                )
+            }
+        };
+        let mut dice_data: oak_dice::evidence::Stage0DiceData =
+            <oak_dice::evidence::Stage0DiceData as zerocopy::FromZeroes>::new_zeroed();
```
[the remark this answers is not in the document]
Comment (jul-sh) [reply oak_restricted_kernel/src/lib.rs:176]: can do! I personally like this form as it clear that a trait is used and which one, but happy to move over for consistency

File: oak_restricted_kernel/src/lib.rs [hunk 5]
```diff
@@ -156,6 +157,45 @@ pub fn start_kernel(info: &BootParams) -> ! {
...
+
+            // Safety: the E820 table indicated that this is the corrct memory segment.
+            unsafe {
+                core::slice::from_raw_parts_mut::<u8>(
+                    start_addr.as_mut_ptr(),
+                    e820_dice_data_entry.size(),
+                )
+            }
+        };
+        let mut dice_data: oak_dice::evidence::Stage0DiceData =
```
Comment (andrisaar) [inline oak_restricted_kernel/src/lib.rs:186]: Would `Stage0DiceData::read_from(dice_memory_slice)` work here?

File: oak_restricted_kernel/src/lib.rs [hunk 6]
```diff
@@ -156,6 +157,45 @@ pub fn start_kernel(info: &BootParams) -> ! {
...
+                .expect("failed to find dice data");
+
+            let start_addr = {
+                let phys_addr = PhysAddr::new(
+                    e820_dice_data_entry
+                        .addr()
+                        .try_into()
+                        .expect("couldn't convert usize to u64"),
+                );
+                VirtAddr::new(phys_addr.as_u64() + mm::KERNEL_OFFSET)
```
Comment (andrisaar) [inline oak_restricted_kernel/src/lib.rs:175]: (a) At this point I think we're still running under identity mapping, so PhysAddr == VirtAddr.
(b) After we've loaded the proper page tables, VirtAddr = PhysAddr + mm:DIRECT_MAPPING_OFFSET.

(See Conrad's comment, below.)

[the remark this answers is not in the document]
Comment (jul-sh) [reply oak_restricted_kernel/src/lib.rs:175]: Good catches, thanks for bearing with me as I wrap my head around this. — Surprisingly the code above does somehow execute, and the resulting dice data is successfully validated.

[the remark this answers is not in the document]
Comment (andrisaar) [reply oak_restricted_kernel/src/lib.rs]: Now that I've refreshed my mind a bit it's not surprising at all :)

See `src/boot/boot.s` -- there's a long comment in there, but tl;dr is that before we set up proper page tables we map the first gigabyte of physical to `mm::KERNEL_OFFSET` in virtual memory as well. It's only after we set up the proper page tables that the region at `mm::KERNEL_OFFSET` should only map memory that's relevant to the kernel executable.

**Review by andrisaar — APPROVED** [review]

**Pull request merged**